packages/foam-vscode/src/features/wikilink-provider.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  Foam,
+  FoamWorkspace,
+  Range,
+  ResourceLink,
+  ResourceParser,
+  URI,
+} from 'foam-core';
+import * as vscode from 'vscode';
+import { FoamFeature } from '../types';
+import { mdDocSelector } from '../utils';
+import { toVsCodeRange, toVsCodeUri } from '../utils/vsc-utils';
+import { createNoteForPlaceholderWikilink } from './create-from-template';
+import { OPEN_COMMAND } from './utility-commands';
+import { monitorFoamVsCodeConfig } from '../services/config';
+
+const GO_TO_DEFINITION_NAV_MODE = 'goToDefinition';
+const OPEN_LINK_NAV_MODE = 'openLink';
+const MIXED_NAV_MODE = 'goToDefinitionButOpenLinkForCreation';
+const NAVIGATION_DISABLED = 'off';
+
+const feature: FoamFeature = {
+  activate: async (
+    context: vscode.ExtensionContext,
+    foamPromise: Promise<Foam>
+  ) => {
+    const foam = await foamPromise;
+
+    const wikilinkProvider = new WikilinkProvider(
+      foam.workspace,
+      foam.services.parser,
+      monitorFoamVsCodeConfig('links.navigationMode')
+    );
+
+    context.subscriptions.push(
+      vscode.languages.registerDefinitionProvider(
+        mdDocSelector,
+        wikilinkProvider
+      ),
+      vscode.languages.registerDocumentLinkProvider(
+        mdDocSelector,
+        wikilinkProvider
+      )
+    );
+  },
+};
+
+export class WikilinkProvider
+  implements vscode.DefinitionProvider, vscode.DocumentLinkProvider {
+  constructor(
+    private workspace: FoamWorkspace,
+    private parser: ResourceParser,
+    private getNavigationMode: () => string
+  ) {}
+
+  public async provideDefinition(
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken
+  ): Promise<vscode.LocationLink[] | vscode.Definition> {
+    const navMode = this.getNavigationMode();
+    if (navMode === OPEN_LINK_NAV_MODE || navMode === NAVIGATION_DISABLED) {
+      return;
+    }
+
+    const resource = this.parser.parse(document.uri, document.getText());
+    const targetLink: ResourceLink | undefined = resource.links.find(link =>
+      Range.containsPosition(link.range, {
+        line: position.line,
+        character: position.character,
+      })
+    );
+
+    if (!targetLink) {
+      return;
+    }
+
+    const uri = this.workspace.resolveLink(resource, targetLink);
+    if (URI.isPlaceholder(uri)) {
+      if (
+        navMode === MIXED_NAV_MODE ||
+        vscode.workspace.workspaceFolders.length < 1
+      ) {
+        return;
+      }
+
+      const newNoteUri = toVsCodeUri(
+        URI.createResourceUriFromPlaceholder(
+          vscode.workspace.workspaceFolders[0].uri,
+          uri
+        )
+      );
+      const newNoteTitle = uri.path.split('/').slice(-1)[0];
+      await createNoteForPlaceholderWikilink(newNoteTitle, newNoteUri);
```

Comment:
For creating a new note from a placeholder I used the same pattern as in `OPEN_COMMAND.execute` (`packages/foam-vscode/src/features/utility-commands.ts`), but only if a VsCode Workspace is open.

Is this the right way to go? @movermeyer  @riccardoferretti

Reply:
My take here is that it's not this provider that should be concerned with the workspace status, but the command itself - as we'd want the behavior to be consistent across "callers"

what's the thinking behind the "exception"?